### PR TITLE
[DDW-663] Use asset policyId+name as unique identifier instead of the fingerprint

### PR DESCRIPTION
**Pull request by daniloprates**
This PR changes the assets identifiers from fingerprint to `policyId+name`.

## Todos

- [x] Remove any fingertip as asset identifier
- [x] Add `policyId+name` as asset identifiers
- [x] Change the types accordingly
- [x] Organize Assets types

## Testing Checklist

- There should be no change in the app behavior
- We need to test anything related to Assets
- [Slack QA thread](https://input-output-rnd.slack.com/messages/GGKFXSKC6)

## Test Cases

```Gherkin
Scenario 1: Displayed data is correctly displayed
Given I am on (Wallet Summary | Native Token Settings) screen
When I hover upon the asset
Then the data is correct
And the data matches the data on production build

Scenario 2: Number of decimal places are correctly displayed
Given I am on Native Token Settings dialog
When I select an option under ‘Number of decimal places’
And I select ‘Save’
Then decimal places are correctly displayed on Wallet Summary screen

Scenario 3: Data displayed as expected on Send screen
Given I am on Send Screen
And I have chosen to send native tokens
When I observe the data
Then the data is correct
And the data matches the data on production build

Scenario 4: Data displayed as expected on Transaction screen
Given I am on Transaction Screen
And I have transaction for (Multiple tokens sent | Multiple tokens received)
And the transaction is expanded
When I hover upon the asset
Then the data is correct
And the data matches the data on production build

Scenario 5: Copying Native Token Data displayed as expected on Transaction screen
Given I am on Transaction Screen
And I have transaction for (Multiple tokens sent | Multiple tokens received)
And the transaction is expanded
And I am hovering upon the asset
When I click on individual data
Then the data is copied
And correct toast message is displayed on top of Daedalus
```
---

### Basics

- [ ] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [ ] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [ ] PR has a good description that summarizes all changes
- [ ] PR has default-sized Daedalus window screenshots or animated GIFs of important UI changes:
  - [ ] In English
  - [ ] In Japanese
- [ ] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [ ] There are no *flow* errors or warnings (`yarn flow:test`)
- [ ] There are no *lint* errors or warnings (`yarn lint`)
- [ ] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [ ] There are no missing translations (running `yarn manage:translations` produces no changes)
- [ ] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [ ] Japanese text changes are proofread and approved (Junko Oda)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [ ] Storybook works and no stories are broken (`yarn storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [ ] Important parts of the code are properly commented and documented
- [ ] Code is properly typed with flow
- [ ] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing
- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

### CODE REVIEW

File: source/renderer/app/api/assets/types.js [hunk 1]
```diff
@@ -1,77 +1,88 @@
...
-  decimals: ?number,
-  recommendedDecimals: ?number,
+export type AssetMetadata = {
+  name: string,
+  ticker: string,
+  description: string,
+  unit?: {
+    decimals: number,
+    name: string,
+  },
```
Comment (rvl) [inline source/renderer/app/api/assets/types.js:83]: To be compatible with the next wallet backend version:

```suggestion
  decimals?: number,
```

File: source/renderer/app/api/api.js [hunk 2]
```diff
@@ -2851,16 +2851,18 @@ const _createAssetFromServerData = action(
       fingerprint,
       metadata,
     } = data;
+    const uniqueId = `${policyId}${assetName}`;
```
Comment (rvl) [inline source/renderer/app/api/api.js:2854]: It's good that this is using the encoded `assetName` and not attempting to decode as ASCII, etc.

**Review by rvl — COMMENTED** [review]
Hi @daniloprates, is the reason for this change that since there is not enough screen space to show the full fingerprint, the AssetId may as well be used?

File: source/renderer/app/api/assets/types.js (hunk 1 above)
Comment (nikolaglumac) [reply source/renderer/app/api/assets/types.js:83]: @daniloprates please keep this as per @rvl's suggestion 🙏

**Comment by nikolaglumac** [comment]
> Hi @daniloprates, is the reason for this change that since there is not enough screen space to show the full fingerprint, the AssetId may as well be used?

No @rvl - this is in order to ensure uniqueness in terms of Daedalus internal asset data handling. So far we treated the fingerprint as the unique identifier which is a false assumption. We are doing this to avoid any future conflicts.

File: source/renderer/app/api/assets/types.js [hunk 3]
```diff
@@ -1,77 +1,88 @@
...
+ * ASSET
+ * Fetched from the Assets API endpoint
+ * It's not attached to a particular wallet or transaction
+ * Therefore, it doesn't have `quantity` nor `address`
+ *
+ * Exclusive data: fingerprint, metadata
+ * Missing data: quantity, address
+ *
+ */
+export type ApiAssets = Array<ApiAsset>;
```
Comment (DominikGuzei) [inline source/renderer/app/api/assets/types.js:17]: Let's move the array type below the single type it is referencing (= move `ApiAssets` definition below the `ApiAsset` definition)

File: source/renderer/app/api/assets/types.js [hunk 4]
```diff
@@ -1,77 +1,88 @@
...
+ */
+export type ApiToken = {
   policy_id: string,
   asset_name: string,
   quantity: number,
+  address?: ?string,
 };
-
-export type WalletAssetItem = {
+export type ApiTokens = Array<ApiToken>;
```
Comment (DominikGuzei) [inline source/renderer/app/api/assets/types.js:50]: same here

File: source/renderer/app/components/wallet/WalletSendForm.js [hunk 5]
```diff
@@ -48,15 +47,11 @@ type Props = {
   currencyMaxIntegerDigits: number,
   currencyMaxFractionalDigits: number,
   currentNumberFormat: string,
-  calculateTransactionFee: (
-    address: string,
-    amount: number,
-    assets: AssetItems
-  ) => Promise<BigNumber>,
+  calculateTransactionFee: Function,
```
Comment (DominikGuzei) [inline source/renderer/app/components/wallet/WalletSendForm.js:50]: is there a reason for making the type signature less specific?

File: source/renderer/app/components/wallet/send-form/WalletSendAssetsConfirmationDialog.js [hunk 6]
```diff
@@ -434,14 +434,14 @@ export default class WalletSendAssetsConfirmationDialog extends Component<
                         assetIndex
                       );
                       return (
-                        <Fragment key={asset.fingerprint}>
+                        <Fragment key={asset.uniqueId}>
```
Comment (DominikGuzei) [inline source/renderer/app/components/wallet/send-form/WalletSendAssetsConfirmationDialog.js:437]: I think this `Fragment` could be removed and the `<div>` below used as the single container element?

File: source/renderer/app/utils/assets.js [hunk 7]
```diff
@@ -28,20 +24,19 @@ export const filterAssets = (
 
 /**
  *
- * This function receives a list os Wallet Assets
- * which don't contain fingerprint nor metadata
- *
- * And a function which gathers this missing items
- * from the `details` in the Assets Store
+ * This function receives a list os Tokens (the assets included in a wallet or transaction)
```
Comment (DominikGuzei) [inline source/renderer/app/utils/assets.js:27]: This comment is misleading / incorrect and seems to be copied from the function below. This function just receives a single token and returns the `AssetToken` from what I can see :)

File: source/renderer/app/stores/AssetsStore.js [hunk 8]
```diff
@@ -56,21 +52,21 @@ export default class AssetsStore extends Store {
...
       const { policyId, assetName } = asset;
-      details[policyId + assetName] = asset;
+      details[`${policyId}${assetName}`] = asset;
       return details;
     }, {});
   }
 
-  getAssetDetails = (policyId: string, assetName: string): ?Asset =>
-    this.details[policyId + assetName];
+  getAssetDomain = (policyId: string, assetName: string): ?Asset =>
```
Comment (DominikGuzei) [inline source/renderer/app/stores/AssetsStore.js:60]: Can you explain a bit what "domain" in `assetDomain` means? I see it in several parts in this PR but don't understand the meaning. Why not `getAsset` instead?

**Review by DominikGuzei — CHANGES_REQUESTED** [review]
Great work 🎉 I just added a few comments 👍

File: source/renderer/app/api/assets/types.js (hunk 1 above)
Comment (daniloprates) [reply source/renderer/app/api/assets/types.js]: ✔️

File: source/renderer/app/api/assets/types.js (hunk 3 above)
Comment (daniloprates) [reply source/renderer/app/api/assets/types.js]: ✔️

File: source/renderer/app/api/assets/types.js (hunk 4 above)
Comment (daniloprates) [reply source/renderer/app/api/assets/types.js:50]: In this case it's already after `ApiToken`, just before `Token`

File: source/renderer/app/components/wallet/WalletSendForm.js (hunk 5 above)
Comment (daniloprates) [reply source/renderer/app/components/wallet/WalletSendForm.js:50]: There is already a specification in the container, WalletSendPage. Having it also inside the component seems duplicated.

File: source/renderer/app/components/wallet/send-form/WalletSendAssetsConfirmationDialog.js (hunk 6 above)
Comment (daniloprates) [reply source/renderer/app/components/wallet/send-form/WalletSendAssetsConfirmationDialog.js:437]: There are 2 divs there, so it needs a Fragment (or `<>`, another `div` or any container).

File: source/renderer/app/utils/assets.js (hunk 7 above)
Comment (daniloprates) [reply source/renderer/app/utils/assets.js]: Nice catch! ✔️

File: source/renderer/app/stores/AssetsStore.js (hunk 8 above)
Comment (daniloprates) [reply source/renderer/app/stores/AssetsStore.js]: I was trying to be more specific, as it was a vague `getDetails`, but it was indeed too much. ✔️

**Review by miorsufianiohk — APPROVED** [review]
LGTM. Tested on 18005 👍

**Review by DominikGuzei — APPROVED** [review]
Code looks great 👍

**Comment by nikolaglumac** [comment]
@daniloprates I will review and merge this tomorrow morning!

**Comment by nikolaglumac** [comment]
@gabriela-ponce please make sure to test this one 🙏

**Comment by gabriela-ponce** [comment]
@daniloprates This looks good, but I find something that I'm not sure if it's expected and I couldn't reproduce it in develop:

- After changing the decimal precision for a Native token, the list refreshes on the "Transactions" screen and the order is altered for a second. Check [this video](https://drive.google.com/file/d/13ZmC2IpvocTVnYpBEu5VdUjkQSNE82bB/view?usp=sharing).

**Comment by daniloprates** [comment]
> @daniloprates This looks good, but I find something that I'm not sure if it's expected and I couldn't reproduce it in develop:
> 
> * After changing the decimal precision for a Native token, the list refreshes on the "Transactions" screen and the order is altered for a second. Check [this video](https://drive.google.com/file/d/13ZmC2IpvocTVnYpBEu5VdUjkQSNE82bB/view?usp=sharing).

Thanks @gabriela-ponce! Fixed.

**Review by gabriela-ponce — APPROVED** [review]

**Review by ManusMcCole — APPROVED** [review]
👍

**Comment by dsc81** [comment]
Hi, 
Is that possible to add the token logo too? Instead of just better naming? 
I think that would be cool features.
Thank you

**Comment by nikolaglumac** [comment]
@dsc81 please submit a feature request here https://github.com/input-output-hk/daedalus/issues
Thanks!

**Pull request merged**